Hibernation: azure actuator

csrwng opened the pull request:
Adds Azure actuator to the hibernation controller

csrwng commented:
Updated to use the ListAll call with statusOnly=true for machine status

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+		return err
+	}
+	if len(machines) == 0 {
+		logger.Warning("No machines were found to stop")
```
Does this deserve to be a warning instead of an info? This could happen if there are machines that are in the Stopping state.

csrwng replied:
The StopMachines call is currently only done when starting to Hibernate. I would think that if you try to hibernate and there's nothing to stop, that's likely not a good thing.

staebler replied:
Oh, so the controller only ever makes one attempt to stop and start the machines. For some reason, I had it in my head that only ever reconcile while stopping, the controller would attempt to stop any machines that are not in the stopped or stopping state.

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+		return err
+	}
+	if len(machines) == 0 {
+		logger.Warning("No machines were found to start")
```
Same question here about whether this should be info level instead of warning. This would happen if there are machines in the Pending state.

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+		logger.Warning("No machines were found to start")
+		return nil
+	}
+	errs := []error{}
```
```suggestion
	var errs []error
```

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+		logger.Warning("No machines were found to stop")
+		return nil
+	}
+	errs := []error{}
```
```suggestion
	var errs []error
```

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+	return len(machines) == 0, nil
+}
+
+type azureMachineLister struct {
```
This is not used.

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+func filterByResourceGroupAndState(machines []compute.VirtualMachine, resourceGroup string, states sets.String, logger log.FieldLogger) []compute.VirtualMachine {
+	result := []compute.VirtualMachine{}
```
```suggestion
	var result []compute.VirtualMachine
```

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+			continue
+		}
+		logger.Infof("Parsed resource: %#v", resource)
+		if strings.ToUpper(resource.ResourceGroup) != strings.ToUpper(resourceGroup) {
```
```suggestion
		if strings.EqualFold(resource.ResourceGroup, resourceGroup) {
```

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+			logger.WithError(err).Warningf("Failed to parse resource ID")
+			continue
+		}
+		logger.Infof("Parsed resource: %#v", resource)
```
This should be debug level (or removed entirely). It outputs a log entry for every machine in the account every time that the machines are listed.

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+			continue
+		}
+		state := azureMachinePowerState(vm)
+		logger.Infof("Machine power state: %s", state)
```
debug level

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+			continue
+		}
+		state := azureMachinePowerState(vm)
+		logger.Infof("Machine power state: %s", state)
+		if !states.Has(state) {
+			continue
+		}
```
Add more context to the log entry.
```suggestion
		logger := logger.WithField("machine", vm.Name).WithField("powerState", state)
		if !states.Has(state) {
			logger.Debug("machine filtered out due to power state"
			continue
		}
		logger.Debug("machine included")
```

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+	if cd.Spec.ClusterMetadata == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s-rg", cd.Spec.ClusterMetadata.InfraID)
```
What is your guess on which OpenShift version changes the name of the resource group and breaks us?

staebler replied:
This is already broken. The OpenShift installer allows the user to specify an existing resource group to use now.
https://github.com/openshift/installer/blob/dbf10881b664be15ba525b9be3562dc478910631/pkg/types/azure/platform.go#L69-L77

staebler replied:
Custom resource groups can be dealt with as part of a fix for https://issues.redhat.com/browse/CO-1150.

staebler commented on pkg/controller/hibernation/azure_actuator.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+func azureMachineNames(machines []compute.VirtualMachine) []string {
+	result := []string{}
```
```suggestion
	var result []string
```

staebler replied:
Or, even better,
```
result := make([]string, len(machines)
for i, m := range machines {
  result[i] = to.String(m.Name)
}
return result
```

staebler commented on pkg/azureclient/client.go:
```diff
@@ -31,6 +31,13 @@ type Client interface {
+
+	// Virtual Machines
+	ListAllVirtualMachines(ctx context.Context, statusOnly string) (compute.VirtualMachineListResultPage, error)
+	ListVirtualMachines(ctx context.Context, resourceGroupName string) (compute.VirtualMachineListResultPage, error)
+	VirtualMachineInstanceView(ctx context.Context, resourceGroupName string, machineName string) (compute.VirtualMachineInstanceView, error)
```
These two functions are not used.

csrwng commented:
@staebler thx for the review. Addressed your comments. Calling start/stop as necessary should be addressed with https://issues.redhat.com/browse/CO-1119

staebler commented on pkg/controller/hibernation/azure_actuator_test.go:
```diff
@@ -0,0 +1,247 @@
+			default:
+				t.Fatal("Invalid function to test")
+			}
+			assert.Nil(t, err)
```
```suggestion
			assert.NoError(t, err)
```

staebler commented on pkg/controller/hibernation/azure_actuator_test.go:
```diff
@@ -0,0 +1,247 @@
+			default:
+				t.Fatal("Invalid function to test")
+			}
+			require.Nil(t, err)
```
```suggestion
			require.NoError(t, err)
```

staebler commented on pkg/controller/hibernation/azure_actuator_test.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+func testAzureClusterDeployment() *hivev1.ClusterDeployment {
+	return testcd.BasicBuilder().Options(func(cd *hivev1.ClusterDeployment) {
+		cd.Spec.ClusterMetadata = &hivev1.ClusterMetadata{
+			InfraID: azureTestInfraID,
+		}
+	}).Build()
```
```suggestion
	return testcd.Build(func(cd *hivev1.ClusterDeployment) {
		cd.Spec.ClusterMetadata = &hivev1.ClusterMetadata{
			InfraID: azureTestInfraID,
		}
	})
```
But there is not much benefit in using a builder here.
```suggestion
	return &hivev1.ClusterDeployment{
		Spec: hivev1.ClusterDeploymentSpec{
			ClusterMetdata: &hivev1.ClusterMetadata{
				InfraID: azureTestInfraID,
			},
		},
	}
```

staebler reviewed (approved):
/lgtm
